src/codewhisperer/service/inlineCompletionService.ts

```diff
@@ -0,0 +1,452 @@
+    private prev?: vscode.Disposable
+    private prevCommand = Commands.declare(
+        'editor.action.inlineSuggest.showPrevious',
+        (context: ExtContext) => async () => {
+            await this.showRecommendation(-1)
+        }
+    )
+    private nextCommand = Commands.declare(
+        'editor.action.inlineSuggest.showNext',
+        (context: ExtContext) => async () => {
+            await this.showRecommendation(1)
+        }
+    )
+    private hideCommand = Commands.declare('editor.action.inlineSuggest.hide', (context: ExtContext) => async () => {
+        await this.clearInlineCompletionStates(vscode.window.activeTextEditor)
+    })
+    constructor() {
+        this.statusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 1)
+        this.documentUri = undefined
+    }
+
+    static #instance: InlineCompletionService
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    registerCommandOverrides() {
+        if (this.extContext) {
+            this.disposeCommandOverrides()
+            this.prev = this.prevCommand.register(this.extContext)
+            this.next = this.nextCommand.register(this.extContext)
+            this.hide = this.hideCommand.register(this.extContext)
+            this.extContext.extensionContext.subscriptions.push(this.prev)
+            this.extContext.extensionContext.subscriptions.push(this.next)
+            this.extContext.extensionContext.subscriptions.push(this.hide)
+        }
+    }
+
+    disposeCommandOverrides() {
+        this.prev?.dispose()
+        this.hide?.dispose()
+        this.next?.dispose()
+    }
+
+    public disposeInlineCompletion() {
+        this.inlineCompletionProviderDisposable?.dispose()
+        this.inlineCompletionProvider = undefined
+    }
+
+    async onEditorChange() {
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+    }
+
+    async onFocusChange() {
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+    }
+
+    async onCursorChange(e: vscode.TextEditorSelectionChangeEvent) {
+        if (e.kind !== 1 && vscode.window.activeTextEditor === e.textEditor) {
+            this.referenceProvider?.removeInlineReference()
+        }
+    }
+
+    async clearInlineCompletionStates(editor: vscode.TextEditor | undefined) {
+        await HoverConfigUtil.instance.restoreHoverConfig()
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+        RecommendationHandler.instance.cancelPaginatedRequest()
+        RecommendationHandler.instance.reportUserDecisionOfCurrentRecommendation(editor, -1)
+        RecommendationHandler.instance.clearRecommendations()
+        this.disposeInlineCompletion()
+        this.setCodeWhispererStatusBarOk()
+        this.disposeCommandOverrides()
+        this.clearRejectionTimer()
```

Comment:
This is _a lot_ of state to keep track of for a single class. Any ideas for removing some of it?

Reply:
As discussed in the description, when inlineCompletionAdditions API(https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.inlineCompletionsAdditions.d.ts#L10) is published, we can remove some states.